# Revisiting Messy Data
So, on a quick preliminary look, I found that the movies Frozen and Tangled looked pretty messy. How messy are they really?

In [1]:
import pandas as pd
import numpy as np

In [2]:
disney = pd.read_csv(r"C:\Users\cassi\Desktop\Data_Science\Animated-Movie-Gendered-Dialogue\private\Disney_1938_2016.csv")

In [3]:
disney.head()

Unnamed: 0 Disney_Period Gender       Movie Role Song Speaker  \
0           0         EARLY    NaN  Snow White  NaN  NaN   queen   
1           1         EARLY    NaN  Snow White  NaN  NaN  mirror   
2           2         EARLY    NaN  Snow White  NaN  NaN   queen   
3           3         EARLY    NaN  Snow White  NaN  NaN  mirror   
4           4         EARLY    NaN  Snow White  NaN  NaN   queen   

  Speaker_Status Start_Scene  \
0          NON-P         NaN   
1          NON-P         NaN   
2          NON-P         NaN   
3          NON-P         NaN   
4          NON-P         NaN   

                                                Text  UTTERANCE_NUMBER  Year  
0  slave in the magic mirror come from the farthe...                 1  1937  
1                what wouldst thou know, my queen ?                  2  1937  
2  magic mirror on the wall, who is the fairest o...                 3  1937  
3  famed is thy beauty, majesty. but hold, a love...                 4  1937  
4                   alas for her ! reveal her name.                  5  1937

In [4]:
disney = disney.drop(columns="Unnamed: 0")

In [93]:
disney.Movie.value_counts()

Frozen                       988
Aladdin                      842
Beauty and the Beast         772
The Princess and the Frog    676
Pocahontas                   638
Mulan                        554
Cinderella                   497
Sleeping Beauty              462
Brave                        411
The Little Mermaid           397
Moana                        382
Snow White                   366
Tangled                      193
Name: Movie, dtype: int64

In [5]:
frozen = disney[disney.Movie == "Frozen"]

In [6]:
frozen.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 988 entries, 5808 to 6795
Data columns (total 11 columns):
Disney_Period       988 non-null object
Gender              0 non-null object
Movie               988 non-null object
Role                0 non-null object
Song                0 non-null object
Speaker             988 non-null object
Speaker_Status      988 non-null object
Start_Scene         0 non-null object
Text                988 non-null object
UTTERANCE_NUMBER    988 non-null int64
Year                988 non-null int64
dtypes: int64(2), object(9)
memory usage: 57.9+ KB


In [47]:
frozen_lines = frozen[["Speaker", "Text"]]

In [48]:
frozen_lines.head()

Speaker                                               Text
5808  ice harvesters   born of cold and winter air and mountain rain ...
5809  ice harvesters   , and his reindeer calf, sven, share a carrot ...
5810  ice harvesters   ice harvesters hup! ho! watch your step! let i...
5811  ice harvesters   ice harvesters stronger than one, stronger tha...
5812  ice harvesters   ice harvesters born of cold and winter air and...

In [49]:
frozen_lines

Speaker                                               Text
5808    ice harvesters   born of cold and winter air and mountain rain ...
5809    ice harvesters   , and his reindeer calf, sven, share a carrot ...
5810    ice harvesters   ice harvesters hup! ho! watch your step! let i...
5811    ice harvesters   ice harvesters stronger than one, stronger tha...
5812    ice harvesters   ice harvesters born of cold and winter air and...
5813    ice harvesters   ice harvesters this icy force both foul and fa...
5814    ice harvesters              ext. the kingdom of arendelle — night 
5815               elsa  sleeps in her bed. her little sister anna: (5)...
5816               anna  elsa. psst. elsa! psst. elsa doesn't stir. ann...
5817               anna                        wake up. wake up. wake up. 
5818               elsa                           anna, go back to sleep. 
5819               anna  i just can't. the sky's awake, so i'm awake, s...
5820               elsa                           ...go play by yourself. 
5821               anna                   do you want to build a snowman? 
5822               anna               come on, come on, come on, come on. 
5823               anna  the girls sneak into the ballroom. elsa shuts ...
5824               anna                       do the magic! do the magic! 
5825               anna                                  this is amazing! 
5826               elsa                                       watch this! 
5827               elsa                hi, i'm olaf and i like warm hugs. 
5828               anna                                 i love you, olaf. 
5829               anna                                         catch me! 
5830               elsa                                           gotcha! 
5831               anna                                     again! again! 
5832               elsa                                        slow down! 
5833               elsa                                       mama! papa! 
5834               elsa  the parents burst through the frozen door. gas...
5835              king   elsa, what have you done? this is getting out ...
5836             queen                                              anna! 
5837               elsa              it was an accident. i'm sorry, anna. 
...                 ...                                                ...
6766              hans   it's a beautiful summer day. the mighty ships ...
6767  french dignitary   i will return this scoundrel to his country. w...
6768               kai                     arendelle thanks you, my lord. 
6769              duke   this is unacceptable. i am innocent. i'm a vic...
6770              duke   ow! my neck hurts. is there a doctor i could.....
6771               kai                   i have a message from the queen. 
6772               kai   arendelle will henceforth and forever no longe...
6773              duke                           weselton. it's weselton! 
6774               anna               come on. come on. come on. come on! 
6775           kristoff                                             pole. 
6776               anna                                      oops. sorry. 
6777               anna                          okay. okay. here we are. 
6778               anna                                 i owe you a sled. 
6779          kristoff                                   are you serious? 
6780               anna                   yes. and it's the latest model. 
6781           kristoff                        no. i can't accept this... 
6782               anna  you have to. no returns. no exchanges. queen's...
6783          kristoff                          what? that's not a thing. 
6784               anna  sure it is. and it even has a cup holder.... d...
6785           kristoff                                          like it? 
6786           kristoff          kristoff i love it.... i could kiss you! 
6787           kristoff  ...i could. i me

In [50]:
frozen_lines.reset_index(inplace=True, drop=True)

In [51]:
frozen_lines

Speaker                                               Text
0      ice harvesters   born of cold and winter air and mountain rain ...
1      ice harvesters   , and his reindeer calf, sven, share a carrot ...
2      ice harvesters   ice harvesters hup! ho! watch your step! let i...
3      ice harvesters   ice harvesters stronger than one, stronger tha...
4      ice harvesters   ice harvesters born of cold and winter air and...
5      ice harvesters   ice harvesters this icy force both foul and fa...
6      ice harvesters              ext. the kingdom of arendelle — night 
7                 elsa  sleeps in her bed. her little sister anna: (5)...
8                 anna  elsa. psst. elsa! psst. elsa doesn't stir. ann...
9                 anna                        wake up. wake up. wake up. 
10                elsa                           anna, go back to sleep. 
11                anna  i just can't. the sky's awake, so i'm awake, s...
12                elsa                           ...go play by yourself. 
13                anna                   do you want to build a snowman? 
14                anna               come on, come on, come on, come on. 
15                anna  the girls sneak into the ballroom. elsa shuts ...
16                anna                       do the magic! do the magic! 
17                anna                                  this is amazing! 
18                elsa                                       watch this! 
19                elsa                hi, i'm olaf and i like warm hugs. 
20                anna                                 i love you, olaf. 
21                anna                                         catch me! 
22                elsa                                           gotcha! 
23                anna                                     again! again! 
24                elsa                                        slow down! 
25                elsa                                       mama! papa! 
26                elsa  the parents burst through the frozen door. gas...
27               king   elsa, what have you done? this is getting out ...
28              queen                                              anna! 
29                elsa              it was an accident. i'm sorry, anna. 
..                 ...                                                ...
958              hans   it's a beautiful summer day. the mighty ships ...
959  french dignitary   i will return this scoundrel to his country. w...
960               kai                     arendelle thanks you, my lord. 
961              duke   this is unacceptable. i am innocent. i'm a vic...
962              duke   ow! my neck hurts. is there a doctor i could.....
963               kai                   i have a message from the queen. 
964               kai   arendelle will henceforth and forever no longe...
965              duke                           weselton. it's weselton! 
966               anna               come on. come on. come on. come on! 
967           kristoff                                             pole. 
968               anna                                      oops. sorry. 
969               anna                          okay. okay. here we are. 
970               anna                                 i owe you a sled. 
971          kristoff                                   are you serious? 
972               anna                   yes. and it's the latest model. 
973           kristoff                        no. i can't accept this... 
974               anna  you have to. no returns. no exchanges. queen's...
975          kristoff                          what? that's not a thing. 
976               anna  sure it is. and it even has a cup holder.... d...
977           kristoff                                          like it? 
978           kristoff          kristoff i love it.... i could kiss you! 
979           kristoff  ...i could. i mean i'd like to. i'd... may i? ...
980               a

In [16]:
frozen_lines.Text.iloc[958] #not a line

"it's a beautiful summer day. the mighty ships have been repaired and are sailing away. on one of the ships, hans is thrown into a brig."

So, a lot of stage commands have been unnecessarily added. 

In [18]:
frozen_lines.Text.iloc[0] #again, lines mixed in with commands

'born of cold and winter air and mountain rain combining, this icy force both foul and fair has a frozen heart worth mining. the men drag giant ice blocks through channels of water. ice harvesters cut through the heart, cold and clear. strike for love and strike for fear. see the beauty sharp and sheer. split the ice apart! and break the frozen heart. hup! ho! watch your step! let it go! '

In [19]:
import re

In [44]:
for line in frozen_lines.head().itertuples():
    print(line[0])
    print(line[1])
    print(line[2])
    print('\n')

0
ice harvesters
born of cold and winter air and mountain rain combining, this icy force both foul and fair has a frozen heart worth mining. the men drag giant ice blocks through channels of water. ice harvesters cut through the heart, cold and clear. strike for love and strike for fear. see the beauty sharp and sheer. split the ice apart! and break the frozen heart. hup! ho! watch your step! let it go!


1
ice harvesters
, and his reindeer calf, sven, share a carrot as they try to keep up with the men.


2
ice harvesters
ice harvesters hup! ho! watch your step! let it go! kristoff struggles to get a block of ice out of the water. he fails, ends up soaked. sven licks his wet cheek. ice harvesters beautiful! powerful! dangerous! cold! ice has a magic can't be controlled.


3
ice harvesters
ice harvesters stronger than one, stronger than ten stronger than a hundred men!


4
ice harvesters
ice harvesters born of cold and winter air and mountain rain combining




In [52]:
frozen_lines.Speaker = frozen_lines.Speaker.map(lambda x: x.strip())

In [53]:
frozen_lines.Text = frozen_lines.Text.map(lambda x: x.strip())

In [37]:
for line in frozen_lines.itertuples():
    if line[2].startswith(line[1]):
        print(line)
        print('\n')

Pandas(Index=2, Speaker='ice harvesters', Text="ice harvesters hup! ho! watch your step! let it go! kristoff struggles to get a block of ice out of the water. he fails, ends up soaked. sven licks his wet cheek. ice harvesters beautiful! powerful! dangerous! cold! ice has a magic can't be controlled.")


Pandas(Index=3, Speaker='ice harvesters', Text='ice harvesters stronger than one, stronger than ten stronger than a hundred men!')


Pandas(Index=4, Speaker='ice harvesters', Text='ice harvesters born of cold and winter air and mountain rain combining')


Pandas(Index=5, Speaker='ice harvesters', Text="ice harvesters this icy force both foul and fair has a frozen heart worth mining. cut through the heart, cold and clear. in the dark, kristoff and sven finally manage to get a single block of ice out of the water. ice harvesters strike for love and strike for fear. there's beauty and there's danger here. split the ice apart! beware the frozen heart.")


Pandas(Index=37, Speaker='bulda', T

In [39]:
#testing .replace()
test = "kristoff is a person kristoff kristoff"
test_new = test.replace('kristoff', 'nerd')
print(test_new)

nerd is a person nerd nerd


In [42]:
test_new_2 = test.replace('kristoff', 'nerd', 1)
print(test_new_2) #only replaces first instance

nerd is a person kristoff kristoff


In [55]:
test_new_2 = test.replace('kristoff ', '', 1)
print(test_new_2) #only replaces first instance

is a person kristoff kristoff


In [56]:
for line in frozen_lines.itertuples():
    if line[2].startswith(line[1]):
        frozen_lines.Text.iloc[line[0]] = line[2].replace(line[1]+' ', '', 1)
        

c:\users\cassi\appdata\local\programs\python\python36-32\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [57]:
frozen_lines.head()

Speaker                                               Text
0  ice harvesters  born of cold and winter air and mountain rain ...
1  ice harvesters  , and his reindeer calf, sven, share a carrot ...
2  ice harvesters  hup! ho! watch your step! let it go! kristoff ...
3  ice harvesters  stronger than one, stronger than ten stronger ...
4  ice harvesters  born of cold and winter air and mountain rain ...

In [58]:
frozen_lines

Speaker                                               Text
0      ice harvesters  born of cold and winter air and mountain rain ...
1      ice harvesters  , and his reindeer calf, sven, share a carrot ...
2      ice harvesters  hup! ho! watch your step! let it go! kristoff ...
3      ice harvesters  stronger than one, stronger than ten stronger ...
4      ice harvesters  born of cold and winter air and mountain rain ...
5      ice harvesters  this icy force both foul and fair has a frozen...
6      ice harvesters              ext. the kingdom of arendelle — night
7                elsa  sleeps in her bed. her little sister anna: (5)...
8                anna  elsa. psst. elsa! psst. elsa doesn't stir. ann...
9                anna                         wake up. wake up. wake up.
10               elsa                            anna, go back to sleep.
11               anna  i just can't. the sky's awake, so i'm awake, s...
12               elsa                            ...go play by yourself.
13               anna                    do you want to build a snowman?
14               anna                come on, come on, come on, come on.
15               anna  the girls sneak into the ballroom. elsa shuts ...
16               anna                        do the magic! do the magic!
17               anna                                   this is amazing!
18               elsa                                        watch this!
19               elsa                 hi, i'm olaf and i like warm hugs.
20               anna                                  i love you, olaf.
21               anna                                          catch me!
22               elsa                                            gotcha!
23               anna                                      again! again!
24               elsa                                         slow down!
25               elsa                                        mama! papa!
26               elsa  the parents burst through the frozen door. gas...
27               king  elsa, what have you done? this is getting out ...
28              queen                                              anna!
29               elsa               it was an accident. i'm sorry, anna.
..                ...                                                ...
958              hans  it's a beautiful summer day. the mighty ships ...
959  french dignitary  i will return this scoundrel to his country. w...
960               kai                     arendelle thanks you, my lord.
961              duke  this is unacceptable. i am innocent. i'm a vic...
962              duke  ow! my neck hurts. is there a doctor i could.....
963               kai                   i have a message from the queen.
964               kai  arendelle will henceforth and forever no longe...
965              duke                           weselton. it's weselton!
966              anna                come on. come on. come on. come on!
967          kristoff                                              pole.
968              anna                                       oops. sorry.
969              anna                           okay. okay. here we are.
970              anna                                  i owe you a sled.
971          kristoff                                   are you serious?
972              anna                    yes. and it's the latest model.
973          kristoff                         no. i can't accept this...
974              anna  you have to. no returns. no exchanges. queen's...
975          kristoff                          what? that's not a thing.
976              anna  sure it is. and it even has a cup holder.... d...
977          kristoff                                           like it?
978          kristoff                    i love it.... i could kiss you!
979          kristoff  ...i could. i mean i'd like to. i'd... may i? ...
980              anna                                            we may.

In [75]:
for line in frozen_lines.itertuples():
    not_line = re.findall(r'(^(the )?[a-z]+s )', line[2])
    if len(not_line) > 0:
        print(line[0], not_line)
        print(line[2])
        print('\n')

5 [('this ', '')]
this icy force both foul and fair has a frozen heart worth mining. cut through the heart, cold and clear. in the dark, kristoff and sven finally manage to get a single block of ice out of the water. ice harvesters strike for love and strike for fear. there's beauty and there's danger here. split the ice apart! beware the frozen heart.


7 [('sleeps ', '')]
sleeps in her bed. her little sister anna: (5) pops up beside her.


15 [('the girls ', 'the ')]
the girls sneak into the ballroom. elsa shuts the door.


17 [('this ', '')]
this is amazing!


26 [('the parents ', 'the ')]
the parents burst through the frozen door. gasp at the sight of the room.


63 [('the gloves ', 'the ')]
the gloves will help.


139 [('princess ', '')]
princess anna of arendelle.


159 [('princess ', '')]
princess anna of arendelle!


358 [('reindeers ', '')]
reindeers are better than people. sven, don't you think that's true? kristoff


380 [('this ', '')]
this is fresh lacquer. seriously, were

So, this approach, to try and match action with words ending is s...not too efficient.

In [76]:
from html.parser import HTMLParser

In [77]:
from bs4 import BeautifulSoup

In [78]:
#okay, so I have these two things. Maybe I can make use of them?

## A Look at Tangled

In [81]:
tangled = disney[disney.Movie == "Tangled"]

In [83]:
tangled_lines = tangled[["Speaker", "Text"]]

In [85]:
tangled_lines.reset_index(drop=True, inplace = True)

In [88]:
tangled_lines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 2 columns):
Speaker    193 non-null object
Text       193 non-null object
dtypes: object(2)
memory usage: 1.5+ KB


Hmmm, only 193 lines. Compared to the other movies, this is already suspicious

In [89]:
for line in tangled_lines.itertuples():
    print(line[1])
    print(line[2])
    print('\n')

flynn
this is the story of how i died. don't worry, this is actually a very fun story and the truth is, it isn't even mine. this is the story of a girl named rapunzel and it starts with the sun now, once upon a time, a single drop of sunlight fell from the heavens and from this small drop of sun, grew a magic, golden flower. it had the ability to heal the sick and injured. oh, you see that old woman over there? you might want to remember her. she's kind of important. well, centuries passed, and a hop skip and a boat ride away there grew a kingdom. the kingdom was ruled by a beloved king and queen. the sick queen and the queen, well, she was about to have a baby. she got sick. really, sick. she was running out of time, and that's when people usually start to look for a miracle. or in this case, a magic golden flower. ah, i told you she'd be important. you see, instead of sharing the sun's gift, this woman, mothe gothel, hoarded its healng power and used it to keep herself for hundreds o

Okay, so this is actually formatted beautifully, but only has the first chunk of the movie! Uh oh.

In [97]:
patf = disney[disney.Movie == "The Princess and the Frog"]

In [98]:
patf_lines = patf[["Speaker", "Text"]]

In [99]:
for line in patf_lines.itertuples():
    print(line[1])
    print(line[2])
    print('\n')

singing
evening star is shining bright, so make a wish and hold on tight, there's magic in the air tonight, and anything can happen. 


eudora
just in that moment, the ugly little frog looked up with his sad, round eyes and pleaded oh, please, dear princess, only a kiss from you can break this terrible spell, that was inflicted on me by a wicked witch. 


charlotte
there comes my favorite part. 


eudora
and the beautiful princess was so moved by his desperate plea, that she stooped down, picked up the slippery creature, leaned forward, raised him to her lips, and kissed that little frog. then the frog was transformed into a handsome prince. they were married and lived happily ever after. the end. 


charlotte
yay! read it again, read it again! 


eudora
sorry charlotte, it's time for us to be headin' home. say good night, tiana. 


tiana
there is no way, in this whole wide world ... i would ever, ever, ever i mean never kiss a frog. yuck! 


charlotte
is that so? here comes your princ



tiana
are you mincing? 


prince naveen
all right ... relax. one! 


tiana
step aside, mister. watch ... and learn. 


prince naveen
all right. 


tiana
there you go. 


prince naveen
you know, i never done anything like this before. 


 tiana
really? 


prince naveen
all right. but, when you are livin' in a castle everything is done for you all the time they dress you, they feed you, drive you... brush your teeth! 


tiana
aw, poor baby. 


 prince naveen
i admitted it was a charmed life, until the day my parents cut me off, and suddenly i realized... i don't know how to do anything. 


 tiana
well hey ... you've got the makings of a decent mushroom mincer! 


 prince naveen
oh, you think so? 


tiana
keep practicing and i just might hire you. 


prince naveen
really? 


tiana
no. 


prince naveen
come on ... what was that?! that was below the frog belt... 


ray
you said hello and he said what. and you said that a ugly fish! 


tiana
anyone for seconds? 


prince naveen
that was ma

## In Summary:
The two movies with the most problems are Frozen and tangled. Both seem to have been scraped from a webpage. While Frozen COULD be edited by going through line by line, this may be tedious. Tangled, on the other hand, is missing the last two-thirds of the movie. I may have to do my own scraping.